### PR TITLE
Bug fix macros.measure with backendv2

### PR DESCRIPTION
**Pull request by to24toro**
### Summary

Currently, we cannot use macros.measure with backendV2, but it is clear that this is an emergency issue that needs to be resolved.
After [the discussion](https://github.com/Qiskit/qiskit-terra/issues/9488#issuecomment-1485382404), the plan is to update macros.measure significantly by adding a new class `MeasureGrouping`. However, this is a major API break issue and will not be ready for the recent 0.25.0 release.
Therefore, as a temporary, I made changes as a version that do not break the API.

### Details and comments
fix https://github.com/Qiskit/qiskit-terra/issues/9488

- Add `meas_map` as a new attribute of `Target`. There is another part of the code that creates an instance of target, however I have not changed it yet now.

- In the `measure` in macors.py, dispatch `measure_v1` and `measure_v2` depending on the class of the backend.

- `FakeHanoiV2` is used as backendV2 in test.

### CODE REVIEW

File: qiskit/pulse/macros.py [hunk 1]
```diff
@@ -11,14 +11,29 @@
...
+    from qiskit.transpiler import Target
+
+
+@deprecate_arg(
+    "backend",
+    deprecation_description=(
+        "Depricating ``backendV1`` as the type of measure's `backend` argument."
+    ),
+    additional_msg=("Instead use ``backendV2``as the type of measure's `backend` argument."),
+    since="0.25.0",
```
Comment (ikkoham) [inline qiskit/pulse/macros.py:34]:
```suggestion
    since="0.24.0",
```
0.25 or 0.24?

File: qiskit/providers/backend_compat.py [hunk 2]
```diff
@@ -56,7 +56,10 @@ def convert_to_target(
     # Parse from properties if it exsits
     if properties is not None:
         qubit_properties = qubit_props_list_from_props(properties=properties)
-        target = Target(num_qubits=configuration.n_qubits, qubit_properties=qubit_properties)
+        target = Target(
```
Comment (nkanazawa1989) [inline qiskit/providers/backend_compat.py:59]: Please revert this change.

File: qiskit/pulse/macros.py [hunk 3]
```diff
@@ -30,6 +45,13 @@ def measure(
     """Return a schedule which measures the requested qubits according to the given
     instruction mapping and measure map, or by using the defaults provided by the backend.
 
+    If the backend has an attribute ``target``, the function uses the measurement logic,
+    "_measure_v2" that takes ``target`` of the ``backend``, ``meas_map`` and ``qubit_mem_slots``
+    assignment.
+    Otherwise, if the backend is None or ``backendV1``, the function uses the
+    measurement logic, "_measure_v1" including ``instruction_schedule_map`` and ``meas_map``
+    as inputs.
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:53]:
```suggestion
    .. note::
        This function internally dispatches schedule generation logic depending on input backend model.
        For the :class:`.BackendV1`, it considers conventional :class:`.InstructionScheduleMap` 
        and utilizes the backend calibration defined for a group of qubits defined in the `meas_map`.
        For the :class:`.BackendV2`, it assembles calibrations of single qubit measurement 
        defined in the backend target to build a composite measurement schedule for `qubits`.
```
This doesn't need to be written here. Note that this is API document that end-users read, so implementation details must be avoided.

File: qiskit/pulse/macros.py [hunk 4]
```diff
@@ -92,6 +163,63 @@ def measure(
...
 
 
+def _measure_v2(
+    qubits: List[int],
+    target: Target,
+    meas_map: Union[List[List[int]], Dict[int, List[int]]],
+    qubit_mem_slots: Dict[int, int],
+    measure_name: str = "measure",
+) -> Schedule:
+    """Return a schedule which measures the requested qubits according to the given
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:173]: 💯

File: qiskit/pulse/macros.py [hunk 5]
```diff
@@ -92,6 +163,63 @@ def measure(
...
+    target and measure map, or by using the defaults provided by the backendV2.
+
+    Args:
+        qubits: List of qubits to be measured.
+        backend (Union[Backend, BaseBackend]): A backend instance, which contains
+            hardware-specific data required for scheduling.
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:179]:
```suggestion
```
This doesn't exist.

File: qiskit/pulse/macros.py [hunk 6]
```diff
@@ -92,6 +163,63 @@ def measure(
...
+                        channels.AcquireChannel(measure_qubit),
+                    ]
+                )
+            else:
+                default_sched = target.get_calibration(measure_name, (measure_qubit,)).filter(
+                    channels=[
+                        channels.AcquireChannel(measure_qubit),
+                    ]
+                )
+        except exceptions.PulseError as ex:
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:212]:
```suggestion
        except KeyError as ex:
```
Target doesn't raise PulseError
https://github.com/Qiskit/qiskit-terra/blob/5128c6751fc2909131ab38c72358bb1e91c9fd84/qiskit/transpiler/target.py#L904-L906

File: qiskit/pulse/macros.py [hunk 7]
```diff
@@ -92,6 +163,63 @@ def measure(
...
+            raise exceptions.PulseError(
+                "We could not find a default measurement schedule called '{}'. "
+                "Please provide another name using the 'measure_name' keyword "
+                "argument. For assistance, the instructions which are defined are: "
+                "{}".format(measure_name, target.instructions)
+            ) from ex
+        schedule += schedule_remapping_memory_slot(default_sched, qubit_mem_slots)
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:219]: 💯

File: qiskit/pulse/macros.py [hunk 8]
```diff
@@ -104,3 +232,33 @@ def measure_all(backend) -> Schedule:
         A schedule corresponding to the inputs provided.
     """
     return measure(qubits=list(range(backend.configuration().n_qubits)), backend=backend)
+
+
+def schedule_remapping_memory_slot(schedule: Schedule, qubit_mem_slots: Dict[int, int]) -> Schedule:
+    """
+    Return a Schedule which is remapped by given qubit_mem_slots.
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:239]:
```suggestion
   A helper function to overwrite MemorySlot index of :class:`.Acquire` instruction.
```

File: qiskit/pulse/macros.py [hunk 9]
```diff
@@ -104,3 +232,33 @@ def measure_all(backend) -> Schedule:
...
+            qubit_index = inst.channel.index
+            reg_index = qubit_mem_slots.get(qubit_index, 0)
+            new_schedule.insert(
+                t0,
+                instructions.Acquire(
+                    inst.duration,
+                    channels.AcquireChannel(qubit_index),
+                    mem_slot=channels.MemorySlot(reg_index),
+                ),
+                inplace=True,
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:260]: 💯

File: qiskit/transpiler/target.py [hunk 10]
```diff
@@ -280,6 +280,7 @@ def __init__(
                 matches the qubit number the properties are defined for. If some
                 qubits don't have properties available you can set that entry to
                 ``None``
+            meas_map (list): List of sets of qubits that must be measured together.
```
Comment (nkanazawa1989) [inline qiskit/transpiler/target.py:283]: Please revert this file

File: qiskit/pulse/macros.py [hunk 11]
```diff
@@ -104,3 +232,33 @@ def measure_all(backend) -> Schedule:
...
+def schedule_remapping_memory_slot(schedule: Schedule, qubit_mem_slots: Dict[int, int]) -> Schedule:
+    """
+    Return a Schedule which is remapped by given qubit_mem_slots.
+
+    Args:
+        schedule: A measurement schedule.
+        qubit_mem_slots: Mapping of measured qubit index to classical bit index.
+
+    Returns:
+        A schedule remapped by qubit_mem_slots as the input provided.
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:246]:
```suggestion
        A measurement schedule with new memory slot index.
```

File: test/python/pulse/test_macros.py [hunk 12]
```diff
@@ -91,6 +92,91 @@ def test_fail_measure(self):
         with self.assertRaises(PulseError):
             macros.measure(qubits=[0], inst_map=self.inst_map)
 
+    def test_measure_v2(self):
+        """Test macro - measure with backendV2."""
+        sched = macros.measure(qubits=[0], backend=self.backend_v2).filter(
```
Comment (nkanazawa1989) [inline test/python/pulse/test_macros.py:97]: Can you add another test with multiple qubits? Also it's better not to filter output (i.e. testing full schedule equality), because some future PR may break this function by injecting invalid instruction. This should be captured by this test.

File: test/python/pulse/test_macros.py [hunk 13]
```diff
@@ -91,6 +92,91 @@ def test_fail_measure(self):
...
+            channels=[MeasureChannel(0), AcquireChannel(0)]
+        )
+        expected = Schedule(
+            (
+                0,
+                Play(
+                    GaussianSquare(
```
Comment (nkanazawa1989) [inline test/python/pulse/test_macros.py:104]: You cannot hard-code these values because backend calibration snapshot may be updated in future.

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]
This is awesome @to24toro ! We can support circuit scheduling with V2 backend with this PR! Proposed implementation looks good, but some minor edits are requested to approve. @mtreinish Can we still add this PR to 0.24 as a bugfix related to BackendV2?

File: qiskit/providers/backend_compat.py (hunk 2 above)
Comment (to24toro) [reply qiskit/providers/backend_compat.py:59]: I fixed it at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/3926200fe62bac8bd9be06c06449bc60d3c93224.

File: qiskit/pulse/macros.py [hunk 14]
```diff
@@ -104,3 +232,33 @@ def measure_all(backend) -> Schedule:
         A schedule corresponding to the inputs provided.
     """
     return measure(qubits=list(range(backend.configuration().n_qubits)), backend=backend)
+
+
+def schedule_remapping_memory_slot(schedule: Schedule, qubit_mem_slots: Dict[int, int]) -> Schedule:
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:237]: @TsafrirA are you happy with this API? This is public so we cannot easily change after release. If you have any concern we can turn this into protected.

File: qiskit/transpiler/target.py (hunk 10 above)
Comment (to24toro) [reply qiskit/transpiler/target.py:283]: I fixed it at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/3926200fe62bac8bd9be06c06449bc60d3c93224.

File: qiskit/pulse/macros.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/pulse/macros.py:34]: TBH, I would just drop the `since` field as this is a pending deprecation it's not on any timer and by default isn't user facing so people knowing that we're planning to deprecate this in the future doesn't really have a start date as there is no timer associated with it yet.

That being said I don't think we need this at all. `BackendV1` is still a support interface in Qiskit we haven't marked it as deprecated or anything yet so I would just remove the use of the deprecation decorator here since I'd expect the pulse builder to support both v1 and v2.

**Review by mtreinish — COMMENTED** [review]
As an isolated bugfix this is fine to include in 0.24.0. Without any API additions it's something we can also merge for 0.24.0 post rc1. I left a couple of small inline comments. Besides that I think a bugfix release release note would be good to document that the schedule is fixed when being passed a backend v2 object.

File: test/python/pulse/test_macros.py (hunk 13 above)
Comment (to24toro) [reply test/python/pulse/test_macros.py]: I add a new test at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/381a8e65d8a4586269ec59ca74a86a01a98661db.

File: test/python/pulse/test_macros.py (hunk 12 above)
Comment (to24toro) [reply test/python/pulse/test_macros.py]: I added a multiple test at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/381a8e65d8a4586269ec59ca74a86a01a98661db.

File: qiskit/pulse/macros.py (hunk 3 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I fixed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/402984987df12b4fc46459db1a57bf8a80d35486.

File: qiskit/pulse/macros.py (hunk 5 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I fixed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/402984987df12b4fc46459db1a57bf8a80d35486.

File: qiskit/pulse/macros.py (hunk 6 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I fixed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/402984987df12b4fc46459db1a57bf8a80d35486.

File: qiskit/pulse/macros.py (hunk 8 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I fixed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/402984987df12b4fc46459db1a57bf8a80d35486.

File: qiskit/pulse/macros.py (hunk 11 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I fixed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/402984987df12b4fc46459db1a57bf8a80d35486.

File: qiskit/pulse/macros.py (hunk 14 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I changed at https://github.com/Qiskit/qiskit-terra/pull/9987/commits/b075cb394e0fed1df63ec1aad5a24b5976fc1c3a.

File: qiskit/pulse/macros.py (hunk 1 above)
Comment (to24toro) [reply qiskit/pulse/macros.py]: I deleted the deprecate decorator because "since" field must be provided.

Comment (nkanazawa1989) [reply qiskit/pulse/macros.py]: @Eric-Arellano here is another case we need to consider :) just fyi

File: qiskit/pulse/macros.py [hunk 15]
```diff
@@ -104,3 +221,35 @@ def measure_all(backend) -> Schedule:
...
+        qubit_mem_slots: Mapping of measured qubit index to classical bit index.
+
+    Returns:
+        A measurement schedule with new memory slot index.
+    """
+    new_schedule = Schedule()
+    for t0, inst in schedule.instructions:
+        if isinstance(inst, instructions.Acquire):
+            qubit_index = inst.channel.index
+            reg_index = qubit_mem_slots.get(qubit_index, 0)
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:243]: The unittest you wrote made me think this should be 
```suggestion
            reg_index = qubit_mem_slots.get(qubit_index, qubit_index)
```
because all acquisition instruction might store results in the slot0 when you specify `qubit_mem_slots={0:0}` (i.e. you may provide the slot only for qubit of interested). This causes data conflict on the slot and you may get meaningless outcome from experiment.

**Review by nkanazawa1989 — CHANGES_REQUESTED** [review]

File: qiskit/pulse/macros.py [hunk 16]
```diff
@@ -92,6 +153,68 @@ def measure(
...
+
+    if isinstance(meas_map, list):
+        meas_map = utils.format_meas_map(meas_map)
+    meas_group = set()
+    for qubit in qubits:
+        meas_group |= set(meas_map[qubit])
+    meas_group = sorted(list(meas_group))
+
+    meas_group_set = set(range(max(meas_group) + 1))
+    unassigned_qubit_indices = list(meas_group_set - qubit_mem_slots.keys())
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:186]:
```suggestion
    unassigned_qubit_indices = list(set(meas_group) - qubit_mem_slots.keys())
```
I think this should work because qubit outside this group doesn't appear.

File: qiskit/pulse/macros.py [hunk 17]
```diff
@@ -92,6 +153,68 @@ def measure(
...
+    unassigned_reg_indices = sorted(list(meas_group_set - set(qubit_mem_slots.values())))
+    if set(qubit_mem_slots.values()).issubset(meas_group_set):
+        for qubit in unassigned_qubit_indices:
+            qubit_mem_slots[qubit] = unassigned_reg_indices.pop(0)
```
Comment (nkanazawa1989) [inline qiskit/pulse/macros.py:190]: This is not blocking but usually pop(0) is slow. For better performance you need to reverse sort `unassigned_reg_indices` and just do `.pop()` here.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @to24toro this looks almost good. Just minor comments. I'm happy to approve once these are addressed.

**Review by nkanazawa1989 — APPROVED** [review]
LGTM

File: qiskit/pulse/macros.py (hunk 1 above)
Comment (Eric-Arellano) [reply qiskit/pulse/macros.py]: Yes, the Sphinx `deprecated` directive expects us to always have a `since`, so that is why it's required.

The docstring for `since` says this:

```
since: The version the deprecation started at. If the deprecation is pending, set
the version to when that started; but later, when switching from pending to
deprecated, update `since` to the new version.
```

> I deleted the deprecate decorator because "since" field must be provided.

I encourage you to still use it and set `pending=True`. That has some benefits:

1. The message will be standardized
2. The pending deprecation will show up in our docs. (It will say its pending, not actually deprecated)
3. It's easy to switch from `pending` to deprecated. You delete `pending=True` and bump `since` to the version it became deprecated.

**Comment by fvoichick** [comment]
I'm relatively inexperienced with Qiskit and I think I'm still running into this issue even after building from source. Could someone please help me or direct me to the appropriate place to ask?

I'm running this code:

    from qiskit import schedule
    from qiskit.circuit import QuantumCircuit
    from qiskit_ibm_provider import IBMProvider

    provider = IBMProvider()
    backend = provider.get_backend('ibmq_quito')
    circuit = QuantumCircuit(1, 1)
    circuit.measure(0, 0)
    schedule(circuit, backend)

I'm getting this error message:

    Traceback (most recent call last):
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/macros.py", line 129, in _measure_v1
    default_sched = inst_map.get(measure_name, measure_group_qubits)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/instruction_schedule_map.py", line 195, in get
    return self._get_calibration_entry(instruction, qubits).get_schedule(*params, **kwparams)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/instruction_schedule_map.py", line 219, in _get_calibration_entry
    self.assert_has(instruction, qubits)
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/instruction_schedule_map.py", line 166, in assert_has
    raise PulseError(
        qiskit.pulse.exceptions.PulseError: "Operation 'measure' exists, but is only defined for qubits [0, 1, 2, 3, 4]."

    The above exception was the direct cause of the following exception:

    Traceback (most recent call last):
    File "/home/finn/Desktop/qiskit/stretch/meas.py", line 9, in <module>
                                                                  schedule(circuit, backend)
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/compiler/scheduler.py", line 101, in schedule
    schedules = parallel_map(schedule_circuit, circuits, (schedule_config, method))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/tools/parallel.py", line 152, in parallel_map
    return [task(values[0], *task_args, **task_kwargs)]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/scheduler/schedule_circuit.py", line 59, in schedule_circuit
    return methods[method](circuit, schedule_config)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/scheduler/methods/basic.py", line 103, in as_late_as_possible
    circ_pulse_defs = lower_gates(circuit, schedule_config)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/scheduler/lowering.py", line 177, in lower_gates
    circ_pulse_defs.append(get_measure_schedule(qubit_mem_slots))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/scheduler/lowering.py", line 98, in get_measure_schedule
    meas_sched = measure(
                 ^^^^^^^^
                 File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/macros.py", line 78, in measure
    return _measure_v1(
        ^^^^^^^^^^^^
        File "/home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages/qiskit/pulse/macros.py", line 131, in _measure_v1
    raise exceptions.PulseError(
        qiskit.pulse.exceptions.PulseError: "We could not find a default measurement schedule called 'measure'. Please provide another name using the 'measure_name' keyword argument. For assistance, the instructions which are defined are: ['sx', 'cx', 'measure', 'id', 'x', 'rz']"

I followed the "Install from source" instructions at https://qiskit.org/documentation/getting_started.html, I'm using the updated `main` branch, and it looks like I'm using the right version:

```
$ pip show qiskit-terra
Name: qiskit-terra
Version: 0.25.0
Summary: Software for developing quantum computing programs
Home-page: https://github.com/Qiskit/qiskit-terra
Author: Qiskit Development Team
Author-email: hello@qiskit.org
License: Apache 2.0
Location: /home/finn/Desktop/qiskit/source/QiskitDevenv/lib/python3.11/site-packages
Requires: dill, numpy, ply, psutil, python-dateutil, rustworkx, scipy, stevedore, symengine, sympy
Required-by: qiskit-ibm-provider
```

This problem was fixed, right? Anyone have ideas about why this still isn't working? Let me know if there's somewhere else I should ask.

**Comment by to24toro** [comment]
@fvoichick
Thank you for asking and sorry for this bug.
Certainly your point is right and we have not fixed scheduling with some backends yet, I will work on some issues related to this bug.
- https://github.com/Qiskit/qiskit-terra/pull/10285
- https://github.com/Qiskit/qiskit-terra/pull/10258
- https://github.com/Qiskit/qiskit-terra/pull/10253
We would appreciate it if you could wait a little longer.
